File: src/main/java/me/jollyperson/featurebot/database/MongoDBDataSource.java
package me.jollyperson.featurebot.database;

import com.mongodb.*;
import com.mongodb.client.*;
import me.jollyperson.featurebot.configuration.Settings;
import me.jollyperson.featurebot.objects.*;
import net.dv8tion.jda.api.entities.Guild;
import org.bson.Document;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static com.mongodb.client.model.Filters.eq;
import static com.mongodb.client.model.Updates.*;

public class MongoDBDataSource implements DatabaseManager{

    private Settings settings;
    private MongoClient mongoClient;
    private MongoDatabase database;
    private MongoCollection<Document> collection;
    private static final Logger logger = LoggerFactory.getLogger(MongoDBDataSource.class);

    @Override
    public void addGuild(BotGuild guild) {
        collection.insertOne(guildToMongo(guild));
    }

    @Override @Nullable
    public BotGuild getGuild(long id) {
        FindIterable<Document> guildResult = collection.find(eq("id", id));
        logger.info(String.valueOf(id));
        GuildSettings guildSettings;
        BotGuild botGuild;
        Set<BotUser> userSet = new HashSet<>();
        List<Document> users;
        for (Document guildMongo : guildResult) {
            Document settingsDoc = guildMongo.get("settings", Document.class);
            guildSettings = new GuildSettings(settingsDoc.getLong("dailyIncrease"),
                    settingsDoc.getBoolean("dailyEnabled"),
                    settingsDoc.getBoolean("logEnabled"),
                    settingsDoc.getLong("logChannelID"),
                    settingsDoc.getString("prefix")
                            );
            users = guildMongo.get("users", List.class);
            users.forEach((document -> {
                //logger.info(document.getLong("id").toString());
                Document eco = document.get("eco", Document.class);
                //logger.info(eco.getLong("balance").toString());
                userSet.add(new BotUser(document.getLong("id"),
                        new BotEconomyUser(
                                eco.getLong("wallet"),
                                eco.getLong("bank"),
                                eco.getLong("debt")),
                        CommandPermission.valueOf(document.getString("permissionLevel"))));
            }));
           // botGuild = new BotGuild(guildMongo.getLong("id"), guildMongo.getString("name"), guildSettings, userSet);
            //logger.info(guildMongo.get("settings", Document.class).getString("prefix"));
            botGuild = new BotGuild(guildMongo.getLong("id"), guildMongo.getString("name"), guildSettings, userSet);
            return botGuild;
        }
        return null;
    }

    private Document guildToMongo(BotGuild botGuild){
        System.out.println(mongoClient);
        GuildSettings settings = botGuild.getGuildSettings();
        Document botGuildMongo = new Document()
                .append("_id", botGuild.getId())
                .append("name", botGuild.getGuildName());
        DBObject guildSettingsMongo = new BasicDBObject();
        guildSettingsMongo.put("logEnabled", settings.isLogEnabled());
        guildSettingsMongo.put("logChannelID", settings.getLogChannelID());
        guildSettingsMongo.put("dailyEnabled", settings.isDailyEnabled());
        guildSettingsMongo.put("dailyIncrease", settings.getDailyEconomy());
        guildSettingsMongo.put("prefix", settings.getPrefix());
        botGuildMongo.put("settings", guildSettingsMongo);

        List<DBObject> usersMongo = new ArrayList<>();
        for (BotUser user : botGuild.getUsers()) {
            DBObject object = new BasicDBObject();
            object.put("id", user.getId());
            object.put("permissionLevel", user.getPerm().toString());
            DBObject eco = new BasicDBObject();
            eco.put("wallet", user.getEco().getWallet());
            eco.put("bank", user.getEco().getWallet());
            eco.put("debt", user.getEco().getWallet());
            eco.put("balance", user.getEco().getWallet());
            object.put("eco", eco);
            usersMongo.add(object);
        }
        botGuildMongo.put("users", usersMongo);
        return botGuildMongo;
    }

    public MongoDBDataSource(Settings settings)  {
        this.settings = settings;
        //mongoClient = new MongoClient(new MongoClientURI("mongodb://" + settings.getMongoDB().getAddress() + ":" + settings.getMongoDB().getPort()));
        mongoClient = MongoClients.create(
                "mongodb://localhost"
        );
        database = mongoClient.getDatabase(settings.getMongoDB().getDatabase());
        collection = database.getCollection(settings.getMongoDB().getCollection());
    }

    @Override
    public boolean isLogging(long guildID) {
        return false;
    }

    @Override
    public void setLogging(long guildID, boolean logging) {

    }

    @Override
    public void setLogID(long guildID, String logChannelID) {

    }

    @Override
    public String getLogID(long guildID) {
        return null;
    }

    @Override
    public String getPrefix(long guildId) {
        return "!";
    }

    @Override
    public void setPrefix(long guildId, String newPrefix) {

        collection.updateOne(
                eq("id",guildId),
                set("settings.prefix", newPrefix));
    }
}

File: src/main/kotlin/me/jollyperson/featurebot/database/MongoDBDataSource.kt
package me.jollyperson.featurebot.database

import com.mongodb.BasicDBObject
import com.mongodb.DBObject
import com.mongodb.MongoClientSettings
import com.mongodb.client.MongoClient
import com.mongodb.client.MongoClients
import com.mongodb.client.MongoCollection
import com.mongodb.client.MongoDatabase
import com.mongodb.client.model.Filters
import com.mongodb.client.model.Updates
import me.jollyperson.featurebot.configuration.Settings
import me.jollyperson.featurebot.objects.*
import okhttp3.internal.addHeaderLenient
import org.bson.Document
import org.slf4j.LoggerFactory
import java.util.*
import java.util.function.Consumer
import kotlin.collections.HashMap

class MongoDBDataSource(private val settings: Settings?) : DatabaseManager {
    private val mongoClient: MongoClient?
    private val database: MongoDatabase
    private val collection: MongoCollection<Document>
    override fun addGuild(guild: BotGuild) {
        collection.insertOne(guildToMongo(guild))
    }

    override fun getGuild(id: Long): BotGuild? {
       /* val guildResult = collection.find(Filters.eq("id", id))
        val guildSettings: GuildSettings
        */val botGuild: BotGuild /*
        val userSet: MutableSet<BotUser?> = HashSet()
        val users: List<Document>
        for (guildMongo in guildResult) {
            val settingsDoc = guildMongo.get("settings", Document::class.java)
            guildSettings = GuildSettings(settingsDoc.getLong("dailyIncrease"),
                    settingsDoc.getBoolean("dailyEnabled"),
                    settingsDoc.getBoolean("logEnabled"),
                    settingsDoc.getLong("logChannelID"),
                    settingsDoc.getString("prefix")
            )
            users = guildMongo.get<List<*>>("users", MutableList::class.java)
            users.forEach(Consumer { document: Document ->
                //logger.info(document.getLong("id").toString());
                val eco = document.get("eco", Document::class.java)
                //logger.info(eco.getLong("balance").toString());
                userSet.add(BotUser(document.getLong("id"),
                        BotEconomyUser(
                                eco.getLong("wallet"),
                                eco.getLong("bank"),
                                eco.getLong("debt")),
                        CommandPermission.valueOf(document.getString("permissionLevel"))))
            })
            // botGuild = new BotGuild(guildMongo.getLong("id"), guildMongo.getString("name"), guildSettings, userSet);
            //logger.info(guildMongo.get("settings", Document.class).getString("prefix"));
            botGuild = BotGuild(guildMongo.getLong("id"), guildMongo.getString("name"), guildSettings, userSet)
            //logger.info(new Gson().toJson(botGuild));

        */
            TODO("fix making it a botguild")
            return botGuild


    }

    private fun guildToMongo(botGuild: BotGuild): Document {
        println(mongoClient)
        val settings = botGuild.guildSettings
        val botGuildMongo = Document()
                .append("id", botGuild.id)
                .append("name", botGuild.guildName)
        val guildSettingsMongo: DBObject = BasicDBObject()
        guildSettingsMongo.put("logEnabled", settings!!.isLogEnabled)
        guildSettingsMongo.put("logChannelID", settings.logChannelID)
        guildSettingsMongo.put("dailyEnabled", settings.isDailyEnabled)
        guildSettingsMongo.put("dailyIncrease", settings.dailyEconomy)
        guildSettingsMongo.put("prefix", settings.prefix)
        botGuildMongo["settings"] = guildSettingsMongo
        val usersMongo: MutableList<DBObject> = ArrayList()
        Document()
                .append("id", botGuild)
                .append("name", botGuild.guildName)
            guildSettingsMongo.put("logEnabled", settings.isLogEnabled)
            guildSettingsMongo.put("logChannelID", settings.logChannelID)
            guildSettingsMongo.put("dailyEnabled", settings.isDailyEnabled)
            guildSettingsMongo.put("dailyIncrease", settings.dailyEconomy)
            guildSettingsMongo.put("prefix", settings.prefix)
            botGuildMongo["settings"] = guildSettingsMongo

        val roles = BasicDBObject()
        botGuild.roles.forEach{
            val group = BasicDBObject()
            group["name"] = it.name
            group["permissions"] = it.permissions
            roles[it.id.toString()] = group
        }
        botGuildMongo["groups"] = roles
        for (user in botGuild.users) {
            val dbject: DBObject = BasicDBObject()
            dbject.put("id", user.id)
            dbject.put("groups", user.groups)
            val eco: DBObject = BasicDBObject()
            eco.put("wallet", user.eco.wallet)
            eco.put("bank", user.eco.wallet)
            eco.put("debt", user.eco.wallet)
            eco.put("balance", user.eco.wallet)
            dbject.put("eco", eco)
            usersMongo.add(dbject)
        }


        botGuildMongo["users"] = usersMongo
        return botGuildMongo
    }

    override fun isLogging(guildID: Long): Boolean {
        return false
    }

    override fun setLogging(guildID: Long, logging: Boolean) {}
    override fun setLogID(guildID: Long, logChannelID: String?) {}
    override fun removeGuild(id: Long) {}
    override fun getStarterWallet(guildID: Long): Long {
        return 0
    }

    override fun getStarterBank(guildID: Long): Long {
        return 0
    }

    override fun getStarterDebt(guildID: Long): Long {
        return 0
    }

    override fun addUser(guildID: Long, user: BotUser?): Boolean {
        return false
    }

    override fun removeUser(guildID: Long, userID: Long): Boolean {
        return false
    }

    override fun hasPermission(guildID: Long, userID: Long, perms: List<PermissionFlag?>?): Boolean {
        return false
    }

    override fun getLogID(guildID: Long): String? {
        return null
    }

    override fun getPrefix(guildId: Long): String? {
        return "!"
    }

    override fun setPrefix(guildId: Long, newPrefix: String?) {
        collection.updateOne(
                Filters.eq("id", guildId),
                Updates.set("settings.prefix", newPrefix))
    }

    companion object {
        private val logger = LoggerFactory.getLogger(MongoDBDataSource::class.java)
    }

    init {
        //mongoClient = new MongoClient(new MongoClientURI("mongodb://" + settings.getMongoDB().getAddress() + ":" + settings.getMongoDB().getPort()));
        mongoClient = MongoClients.create(MongoClientSettings.builder().build())
        if (mongoClient == null) {
            logger.info("NULLLLLELRLTR")
        }
        database = mongoClient.getDatabase(settings.getMongoDB().database)
        collection = database.getCollection(settings.getMongoDB().collection)
    }
}